**FAVITES** (FrAmework for VIral Transmission and Evolution Simulation) is a robust modular framework for the simultaneous simulation of a transmission network and viral evolution, as well as simulation of sampling imperfections of the transmission network and of the sequencing process ([Moshiri *et al*., 2018](https://doi.org/10.1093/bioinformatics/bty921)). The framework is robust in that the simulation process has been broken down into a series of interactions between abstract module classes, and the user can simply plug in each desired module implementation (or implement one from scratch) to customize any stage of the simulation process.

Because of the modular nature of FAVITES, a single README would become too convoluted. As a result, all information, including installation information, requirements, usage, file format descriptions, etc., are located in the [FAVITES Wiki](https://github.com/niemasd/FAVITES/wiki).

Note that many FAVITES module implementations wrap around existing tools, so in addition to citing FAVITES, you should cite the tools wrapped in any module implementations you use in your study.

# Citing FAVITES
If you use FAVITES in your work, please cite:

> **Moshiri N**, Ragonnet-Cronin M, Wertheim JO, Mirarab S (2018). "FAVITES: simultaneous simulation of transmission networks, phylogenetic trees, and sequences." *Bioinformatics*. 35(11):1852-1861. [doi:10.1093/bioinformatics/bty921](https://doi.org/10.1093/bioinformatics/bty921)
